In [89]:
from pathlib import Path

data_path = Path("../data/raw/archive (1)/nyai-supreme-court-clean")

files = list(data_path.rglob("*.md"))

print("Total Files:", len(files))

Total Files: 33016


## Checking Data Consistency

To evaluate the feasibility of automated metadata extraction, random judgments were sampled across multiple decades and inspected manually.

In [90]:
import random

samples = random.sample(files, 10)

for file in samples:
    print("=" * 100)
    print(file.name)

    with open(file, encoding="utf-8") as f:
        text = f.read()

    print(text[:1500])

2014_8_369_374_EN.md
[2014] 8 S.C.R. 369

MOHINDER SINGH & ORS.

STATE OF HARYANA (Civil Appeal Nos.7227-7257 of 2014 etc).

AUGUST 05, 2014

[T.S. THAKUR, C. NAGAPPAN AND

ADARSH KUMAR GOEL, JJ.]

Land Acquisition Act, 1894 - Land acquired under - Award

of compensation - Reasonable deduction from market value C of the land, towards development charges - Ascertainment of

- Reference Court deducted 1/41h of the market value - High Court deducted 40% of the market value -Held: If the acquired land is already in the midst of already developed land, deduction of 40% would not be justified - In the present case, D the acquired land was already within developed municipal limits, hence the cut of 40% is not justified - Deduction of

1141h of market .value made by Reference Court is appropriate.

The question for consideration in the present E appeals was as to what would be the reasonable deduction towards development charges, to be made from the market value, in respect of the land in ques

### Observations

- Judgment headers follow a largely consistent structure across different years.
- Key metadata fields such as case name, judgment date, and presiding judges appear near the beginning of most documents.
- Legal topics and relevant acts are often explicitly mentioned in the opening sections.
- Minor formatting variations exist, particularly in older OCR-processed documents, but the overall structure remains consistent.

### Conclusion

The dataset demonstrates a high degree of structural consistency, suggesting that rule-based metadata extraction using regular expressions and text processing techniques is likely to achieve high extraction accuracy.

In [91]:
import re

date_patterns = [
    r"[A-Z]+\s+\d{1,2},\s+\d{4}",
    r"[A-Z][a-z]+\s+\d{1,2},\s+\d{4}",
    r"\d{1,2}\s+[A-Z][a-z]+\s+\d{4}"
]



In [92]:
import random

samples = random.sample(files, 10)

failed = []

for file in samples:
    print(extract_year(file))

    with open(file, encoding="utf-8") as f:
        text = f.read()
    
    found = False

    for date_pattern in date_patterns:
        match = re.search(date_pattern, text)
        if match:
            print(match.group())
            found = True
            break

        if not found:
            failed.append(file)
    
    print("=" * 50)

for file in failed:
    with open (file, encoding="utf-8") as f:
        text = f.read()
    print (text[:1500])
    print ("=" * 50)

2016
August 23, 2012
1994
MARCH 18, 1994
2022
NOVEMBER 17, 2022
2009
FEBRUARY 12, 2009
2006
SEPTEMBER 29, 2006
1954
January 24, 1950
1979
May 2, 1979
2021
SEPTEMBER 24, 2021
2025
21 February 2025
1954
[2016] 4 S.C.R. 816

STATE OF GUJARAT & ANR.

LAL SINGH @ MANJIT SINGH & ORS.

(Criminal Appeal No. 171 of2016)

JUNE29,2016

[DIPAK MISRA AND SHIVA KffiTI SINGH, JJ.]

Code of Criminal Procedure, 1973 - ss. 432 and 433 - Remission under - Respondent was convicted under the provisions of TADA by .the Designated Court in the State of Gujarat and sentenced to life imprisonment - Conviction was confirmed by Supreme Court - On the request of the respondent-convict, he was transferred from the jail in Gujarat to the jail in Punjab -Application by the convict for remission of sentence - Rejected by the State of Gujarat - Writ petition challenging the order of refusal - High Court directed the State of Gujarat to reconsider the representation of the convict to release him premature - On appeal, 

Initial metadata extraction experiments indicate that judgment dates can be extracted with high success rates using a small set of rule-based patterns. Remaining failures are primarily attributable to OCR artifacts and historical formatting variations rather than structural inconsistencies in the dataset.

In [93]:
for file in random.sample(files, 10):

    with open(file, encoding="utf-8") as f:
        text = f.read()

    lines = text.split("\n")

    for i, line in enumerate(lines[:30]):

        if "v." in line.lower():
            print(file.name)
            print(lines[max(0,i-2):i+3])
            print("="*50)
            break

1989_3_670_679_EN.md
['STATE OF RAJASTHAN & ANR. v.', '', 'SHANTILAL ETC.']
2001_2_797_806_EN.md
['NEW INDIA ASSURANCE CO., SHIMLA A', '', '~ v.', '', 'KAMLA AND ORS. ETC. ETC.']
S_1985_2_879_897_EN.md
['SEPrEM!lER 20, 1985.', '', '[V.D. TUUAPURKAR AND SABYASACHI MUKHAR.Jl, JJ.]', '', 'Income Tax Act, 1922, s.25(4) - Hindu undivided family - carrying on business - Assessed under Act of 191~ - Partial partition of family on 11th October, 1948 - Partnership firm succeeding family business on 12th October, 1948 - Succession - When takes place - Intention to carry on business Relevancy of']
1995_3_544_548_EN.md
['The appellants holding only diploma qualilications were appointed as Assistant Inspector (Factories) in 1963, because persons with pr.!scribed minimum qualifications viz. degree in Engineering were not available. Subsequently, when Mysore State Government Service (Recruit- E ment of Local Candidates to Class III Posts) Rules, 1966 and Mysore Labour Service (Recruitment) Rules, 196

In [94]:
def extract_case_name(text):

    lines = [line.strip() for line in text.split("\n") if line.strip()]

    for i, line in enumerate(lines[:40]):

        if line.lower() == "v.":
            return f"{lines[i-1]} v. {lines[i+1]}"

        if " v." in line.lower():
            return line

    return None

In [95]:

for file in random.sample(files, 100):
    with open(file, encoding="utf-8") as f:
        text = f.read()
    print(extract_case_name(text))

THE ENTERTAINMENT TAX OFFICER A v.
Division Bench of this Court doubting the correctness of the decision in Ganpat Roy and Ors. v. Addi. Distt. Magistrate and Ors., (1985) 2 SCC p 307 referred the matter to larger Bench. Appellant also filed two interlocutory applications seeking direction to the respondents to pay the rent in arrears and for direction to respondents to vacate the building.
None
STATE OF U.P. & ANR. v. U.P. RAJYA KHANIJ VIKAS 537
STATE OF UTTARANCHAL & ORS. v. 55
SWATI ULHAS KERKAR AND ORS. v. SANJAY
The appellant's v..Tit petition challenging the Governments nction in taking disciplinary action against him after retirement was dismissed by the High Court.
AMARJIT KAUR v. KARAMVIR SINGH [PASA Y AT, J.] 229
G. S. REDDI v. GOVT. OF A.P. 485
POOJA BATRA A v.
None
A v.
(S.B. SINHA AND V.S. SIRPURKAR, JJ.) .., l..
A v.
None
HATIISINGH A v.
CARGILL INDIA PVT. LTD. v. COMMNR. OF CUSTOMS & 919
VINAY KUMAR RAI AND ANR. '<.. v.
None
Gurcharan Singh and Anr. v. State of Punjab, A

In [96]:
def extract_header(text, n_lines=50):

    lines = text.split("\n")

    return "\n".join(lines[:n_lines])

In [97]:
def extract_body(text, n_lines=50):

    lines = text.split("\n")

    return "\n".join(lines[n_lines:])

In [98]:
import random

file = random.choice(files)

with open(file, encoding="utf-8") as f:
    text = f.read()

header = extract_header(text)
body = extract_body(text)

print("HEADER")
print("=" * 50)
print(header[:1500])

print("\nBODY")
print("=" * 50)
print(body[:1500])

HEADER
GRINDLA YS BANK LIMITED

THE INCOME TAX OFFICER, 'H' WARD COMPANIES,

DISTRICT-IV, CALCUTTA AND ORS.

January 15, 1980

[N. L. UNTWALIA AND R. s. PATHAK, JJ.]

Bar of linlitation under section 153(1)(a)(iii) of the lnco111e Tax Act, 1961··- When the asssment proceeding remained during the entire per!od by successive orders of the Court, the fresh assessn1ent order cannot be faulted 011 grounds of limitation.

Powers of High Court to make the order a ftehll a, sessnient und,::r certiorari ; urisdiction under Art. 226 of the Constitution.

The appellant, a banking compooy incorporated in the United Kingdom, carries on banking business in India and is assessed under the Income Ta-x Art,

1961. The appellant filed a return of it~ income for the assessment year 1972-73.

During the assessment proceedings the1 Income Tax Officer issued a notice under •cction 142 (I) of the Income Ta, x Act requiring the appellant to produce certain account books and documents. The appellant applied ag

### Document Segmentation

A simple line-based segmentation strategy was tested by treating the first 50 lines of each judgment as the document header and the remaining text as the body.

The header consistently contains high-value metadata including:
- Case title
- Date of judgment
- Judges
- Jurisdiction
- Appeal information

The body contains:
- Facts of the case
- Legal arguments
- Statutory references
- Judicial reasoning

This suggests that metadata extraction and content analysis should be treated as separate processing stages.

In [99]:
def extract_year(file):

    parts = file.stem.split("_")

    if parts[0].isdigit():
        return int(parts[0])

    return int(parts[1])

In [100]:
import re

date_patterns = [
    r"[A-Z]+\s+\d{1,2},\s+\d{4}",
    r"[A-Z][a-z]+\s+\d{1,2},\s+\d{4}",
    r"\d{1,2}\s+[A-Z][a-z]+\s+\d{4}",
    r"[A-Z]+\d{1,2},\d{4}"
]

In [101]:
def extract_date(header):

    for pattern in date_patterns:

        match = re.search(pattern, header)

        if match:
            return match.group()

    return None

In [102]:
def extract_judges(header):

    match = re.search(r"\[(.*?)\]", header, re.DOTALL)

    if match:
        return match.group(1)

    return None

In [103]:
def extract_case_type(header):

    lines = header.split("\n")

    for line in lines:

        if "JURISDICTION" in line.upper():
            return line.strip()

    return None

In [104]:
def extract_metadata(file):

    with open(file, encoding="utf-8") as f:
        text = f.read()

    header = extract_header(text)

    metadata = {
        "file": file.name,
        "year": extract_year(file),
        "date": extract_date(header),
        "judges": extract_judges(header),
        "case_type": extract_case_type(header)
    }

    return metadata

In [105]:
sample = extract_metadata(files[0])

sample

{'file': '1950_1_1008_1018_EN.md',
 'year': 1950,
 'date': None,
 'judges': '1950',
 'case_type': 'The jurisdiction of the High Court in the matter of incometax references is an advisory jurisdiction and under the Inco:netax Act the decision of the Appellate Tribunal on facts is final unless it can be successfully assailed on the ground tha~ there was'}

In [106]:
metadata = []

for file in random.sample(files, 100):

    metadata.append(
        extract_metadata(file)
    )

In [107]:
import pandas as pd

metadata_df = pd.DataFrame(metadata)

metadata_df.head()

,file,year,date,judges,case_type
0,S_1975_1_265_269_EN.md,1975,"April 23, 1975","V, .R. KRISHNA IYER, R. S. SARKARIA AND A. C. ...",NaN
1,2009_1_711_722_EN.md,2009,"FEBRUARY 2, 2009","DR. ARIJIT PASAYAT, P. SATHASIVAM AND\n\nAFTAB...",CIVIL APPELLATE JURISDICTION: Civil Appeal No....
2,S_1993_3_694_699_EN.md,1993,"NOVEMBER 16, 1993","M.N. VENKATACHALIAH, CJI, S. MOHAN AND\n\nDR. ...",B CIVIL APPELLATE JURISDICTION: Civil Appeal N...
3,1970_3_662_679_EN.md,1970,"February 12, 1970",668 B-E,NaN
4,S_2000_2_302_309_EN.md,2000,"AUGUST 9, 2000","S.P. BHARUCHA, R.C. LAHOTI AND N. SANTOSH HEGD...",CIVIL APPELLATE JURISDICTION : Civil Appeal No...


In [108]:
metadata_df.isna().sum()

file          0
year          0
date          7
judges        3
case_type    43
dtype: int64

In [109]:
metadata_df.notna().mean()

file         1.00
year         1.00
date         0.93
judges       0.97
case_type    0.57
dtype: float64

In [110]:
failed_case_types = metadata_df[
    metadata_df["case_type"].isna()
]

failed_case_types["file"].head(10)

0      S_1975_1_265_269_EN.md
3        1970_3_662_679_EN.md
5        1986_2_621_699_EN.md
10    2020_13_1044_1073_EN.md
23      2020_13_925_950_EN.md
25      2020_10_579_595_EN.md
28       1960_3_810_819_EN.md
30       2020_4_770_793_EN.md
32     2017_5_1005_1018_EN.md
33    2019_13_1028_1043_EN.md
Name: file, dtype: str

In [111]:
failed_case_types.iloc[1]

file         1970_3_662_679_EN.md
year                         1970
date            February 12, 1970
judges                    668 B-E
case_type                     NaN
Name: 3, dtype: object

In [112]:
for file in files:

    if file.name == "2012_9_579_600_EN.md":

        with open(file, encoding="utf-8") as f:
            text = f.read()

        print(extract_header(text))
        break

[2012] 9 S.C.R. 579

PUBLIC UNION FOR CIVIL LIBERTIES

STATE OF TAMIL NADU & ORS. (Writ Petition (Civil) No. 3922 of 1985)

OCTOBER 15, 2012

[K.S. RADHAKRISHNAN AND DIPAK MISRA, JJ.)

BONDED LABOUR SYSTEM (ABOLITION) ACT, 1976:

ss. 10, 11 and 12 - Bonded labour - Abolition of - c Rehabilitation of freed bonded labourers - Directions issued to States/Union Territories to conduct periodical surveys in accordance with provisions of the Act, to calculate firm requirement of funds for rehabilitation of freed bonded labourers and to take steps to enhance the rehabilitation 0 package - States and UTs should continue. to submit six monthly reports to NHRC and the latter would effectively supervise and take appropriate steps for carrying out the provisions of the Act and directions issued by the Court - Right of Children to Free and Compulsory Education Act, 2009 - Minimum Wages Act, 1948, Workmen's Compensation Act, E 1923 - Inter State Migrant Workmen Act, 1979 - Child Labour (Prohibition a

## Conclusions

- Dataset contains 33,016 Supreme Court judgments.
- Judgment structure is largely consistent across decades.
- Important metadata is concentrated near the beginning of documents.
- Header/body segmentation is effective for separating metadata from legal reasoning.
- Initial metadata extraction results:

    - Year: 100%
    - Date: 93%
    - Judges: 97%
    - Case Type: 57%

These findings support the feasibility of constructing an automated legal intelligence pipeline for downstream tasks such as metadata indexing, semantic search, clustering, anomaly detection, and legal analytics.